notte/common/parser.py

```diff
@@ -62,69 +79,128 @@ def search_pattern(text: str, tag: str) -> str | None:
         return match.group(1).strip() if match else None
 
     @override
-    def which(self, text: str) -> Literal["observe", "step", "rules"]:
-        url = self.search_pattern(text, "url")
-        action = self.search_pattern(text, "action")
-        match (bool(url), bool(action)):
-            case (True, False):
+    def which(self, text: str) -> Literal["observe", "step", "scrape", "rules"]:
+        url = self.search_pattern(text, BaseNotteParser.observe_tag)
+        action = self.search_pattern(text, BaseNotteParser.step_tag)
+        scrape = self.search_pattern(text, BaseNotteParser.scrape_data_tag)
+        match (bool(url), bool(action), bool(scrape)):
+            case (True, False, False):
                 return "observe"
-            case (False, True):
+            case (False, True, False):
                 return "step"
+            case (False, False, True):
+                return "scrape"
             case _:
                 return "rules"
 
+    @override
+    def is_done(self, text: str) -> bool:
+        return (
+            f"<{BaseNotteParser.done_tag}/>" in text or self.search_pattern(text, BaseNotteParser.done_tag) is not None
+        )
+
+    @override
+    def get_done_answer(self, text: str) -> str | None:
+        if not self.is_done(text):
+            raise ValueError("Not done")
+        return self.search_pattern(text, BaseNotteParser.done_tag)
+
     @override
     def observe(self, text: str) -> EnvObserveParams:
-        url = self.search_pattern(text, "url")
+        url = self.search_pattern(text, BaseNotteParser.observe_tag)
         if url is None:
             raise ValueError("No URL found")
         return EnvObserveParams(url=url)
 
     @override
     def step(self, text: str) -> EnvStepParams:
-        action = self.search_pattern(text, "action")
+        action = self.search_pattern(text, BaseNotteParser.step_tag)
         if action is None or not isinstance(action, str):
             raise ValueError("No action found")
-        d = json.loads(action)
-        if "action-id" not in d:
+        try:
+            action_dict: ActionJson = json.loads(action)
+        except json.JSONDecodeError:
+            raise ValueError("Invalid JSON in action")
+        action_id = action_dict.get("action_id", None)
+        if action_id is None:
             raise ValueError("No action-id found in action")
-        action_id = d["action-id"]
-        params = d.get("params", None)
+        params = action_dict.get("params", None)
         return EnvStepParams(action_id=action_id, params=params)
 
     @override
     def rules(self) -> str:
         return self.INSTRUCTIONS
 
-    @override
-    def textify(self, obs: Observation) -> str:
-        if obs.space is None:
+    def final_rules(self) -> str:
```

Comment:
We have a variable `INSTRUCTIONS` that you can rename `PRE_INSTRUCTIONS` and move this function to another variable called `POST_INSTRUCTIONS` it makes more sense to me